File: src/mvmctl/core/image/_provisioner.py
"""
ImageProvisioner — image optimization via backends.

Debloats (OS cache cleanup + fstab fix) and optionally shrinks a root
filesystem image, then executes them via the selected backend (loop-mount
or guestfs).  Debloat and shrink are run as **separate backend sessions**
so a shrink failure (e.g. filesystem already at minimum size) never
prevents the debloat step from being applied::

    from mvmctl.core.image._provisioner import ImageProvisioner

    p = ImageProvisioner(
        image_path=...,
        provisioner_type=ProvisionerType.LOOP_MOUNT,
        fs_type="ext4",
    )
    p.deblob()
    p.shrink()
    p.run()
"""

from __future__ import annotations

import logging
from pathlib import Path

from mvmctl.core._shared._provisioner._backend import ProvisionerBackend
from mvmctl.models.provisioner import ProvisionerType

logger = logging.getLogger(__name__)


class ImageProvisioner:
    """Optimize a root filesystem image — shrink, deblob, fix fstab.

    ``deblob()`` and ``shrink()`` are **declarative** — they only set
    flags.  ``run()`` creates a fresh backend for each phase so a
    failure in one phase never leaks into the next.
    """

    def __init__(
        self,
        image_path: Path,
        *,
        provisioner_type: ProvisionerType,
        fs_type: str,
    ) -> None:
        self._image_path = image_path
        self._provisioner_type = provisioner_type
        self._fs_type = fs_type
        self._deblob = False
        self._shrink = False
        self._convert_to: str | None = None

    # -- builder methods (declarative) ------------------------------------

    def detect_os(self) -> str:
        """Detect the OS type from the image using a fresh backend session.

        Returns:
            OS identifier string (e.g. ``"ubuntu"``, ``"debian"``, ``"alpine"``).

        """
        backend = ProvisionerBackend.get_image(
            self._image_path,
            provisioner_type=self._provisioner_type,
            fs_type=self._fs_type,
        )
        return backend.detect_os()

    def deblob(self) -> None:
        """Mark that deblob + fstab fix should run."""
        self._deblob = True

    def shrink(self) -> None:
        """Mark that filesystem shrink should run."""
        self._shrink = True

    def convert_to(self, target_fs: str) -> None:
        """Mark that filesystem conversion should run as Phase 0.

        Useful for converting btrfs images to ext4 before deblob/shrink.
        The conversion runs **first** (Phase 0) so that all subsequent
        operations see the converted filesystem.
        """
        self._convert_to = target_fs

    # -- execution ---------------------------------------------------------

    def run(self) -> bool:
        """Execute queued operations with the selected backend.

        Phases run in order — conversion (Phase 0), deblob (Phase 1),
        shrink (Phase 2).  Each phase uses a **fresh backend session** so a
        failure in one phase never leaks into the next.

        * Phase 0: filesystem conversion (e.g. btrfs → ext4).  Runs first
          so that deblob and shrink see the converted filesystem.
        * Phase 1: deblob + fstab fix.
        * Phase 2: filesystem shrink (ext-family only).

        Returns:
            True if at least one phase ran successfully, False if all
            phases were skipped or failed.

        """
        from mvmctl.exceptions import LoopMountError

        deblob_ok = False
        shrink_ok = False
        convert_ok = False

        # Phase 0: filesystem conversion (e.g. btrfs → ext4)
        # Runs first so that deblob and shrink operate on the converted fs.
        if self._convert_to is not None:
            backend = ProvisionerBackend.get_image(
                self._image_path,
                provisioner_type=self._provisioner_type,
                fs_type=self._fs_type,
            )
            try:
                backend.convert_to(self._convert_to)
                self._fs_type = self._convert_to
                convert_ok = True
                logger.info(
                    "Filesystem converted: %s → %s",
                    self._fs_type,
                    self._convert_to,
                )
            except (LoopMountError, OSError, RuntimeError) as exc:
                logger.warning(
                    "Filesystem conversion skipped: %s. "
                    "Build the provisioner binary with 'python scripts/build_services.py' "
                    "or enable libguestfs to enable fs conversion.",
                    exc,
                )

        # Phase 1: deblob + fstab fix (fresh backend — no state leakage)
        if self._deblob:
            backend = ProvisionerBackend.get_image(
                self._image_path,
                provisioner_type=self._provisioner_type,
                fs_type=self._fs_type,
            )
            backend.deblob()
            try:
                backend.run()
                deblob_ok = True
            except (LoopMountError, OSError, RuntimeError) as exc:
                logger.warning(
                    "Debloating skipped: %s. "
                    "Build the provisioner binary with 'python scripts/build_services.py' "
                    "or enable libguestfs to enable boot optimization.",
                    exc,
                )

        # Phase 2: shrink (fresh backend — deblob state completely isolated)
        if self._shrink:
            backend = ProvisionerBackend.get_image(
                self._image_path,
                provisioner_type=self._provisioner_type,
                fs_type=self._fs_type,
            )
            backend.shrink()
            try:
                backend.run()
                shrink_ok = True
            except (LoopMountError, OSError, RuntimeError) as exc:
                logger.debug(
                    "Shrink skipped (image may already be minimal): %s", exc
                )

        return convert_ok or deblob_ok or shrink_ok

File: internal/core/image/provisioner.go
package image

import (
	"context"
	"errors"
	"fmt"
	"log/slog"

	"mvmctl/internal/infra"
	"mvmctl/internal/lib/provisioner"
	"mvmctl/pkg/errs"
)

// Provisioner optimizes a root filesystem image — shrink, deblob, fix fstab.
// deblob() and shrink() are declarative — they only set flags.
// run() creates a fresh backend for each phase.
type Provisioner struct {
	imagePath       string
	provisionerType provisioner.ProvisionerType
	fsType          string
	deblob          bool
	shrink          bool
	convertTo       string
}

// NewProvisioner creates a new Provisioner.
func NewProvisioner(
	imagePath string,
	provisionerType provisioner.ProvisionerType,
	fsType string,
) *Provisioner {
	return &Provisioner{
		imagePath:       imagePath,
		provisionerType: provisionerType,
		fsType:          fsType,
	}
}

// createBackend creates a fresh backend for the current image.
func (p *Provisioner) createBackend(ctx context.Context) (provisioner.Backend, error) {
	cacheDir, err := infra.GetCacheDir()
	if err != nil {
		return nil, fmt.Errorf("resolve cache dir: %w", err)
	}
	return provisioner.NewBackend(ctx, provisioner.BackendOpts{
		RootfsPath:      p.imagePath,
		FsType:          p.fsType,
		CacheDir:        cacheDir,
		ProvisionerType: p.provisionerType,
		UserUID:         1000,
		UserGID:         1000,
	})
}

// --- builder methods (declarative) ---

// DetectOS detects the OS type from the image using a fresh backend session.
// Returns OS identifier (e.g. "ubuntu", "debian", "alpine").
// Errors propagate to the caller.
func (p *Provisioner) DetectOS(ctx context.Context) (string, error) {
	backend, err := p.createBackend(ctx)
	if err != nil {
		return "", err
	}
	return backend.DetectOS(ctx)
}

// Deblob marks that deblob + fstab fix should run.
func (p *Provisioner) Deblob() {
	p.deblob = true
}

// Shrink marks that filesystem shrink should run.
func (p *Provisioner) Shrink() {
	p.shrink = true
}

// ConvertTo marks that filesystem conversion should run as Phase 0.
func (p *Provisioner) ConvertTo(targetFS string) {
	p.convertTo = targetFS
}

// --- execution ---

// Run executes queued operations with the selected backend.
// Phases run in order — conversion (Phase 0), deblob (Phase 1), shrink (Phase 2).
// A single backend session is shared across all phases — unnecessary
// mount/umount cycles are avoided.  ConvertTo updates the backend's internal
// fsType so subsequent phases see the converted filesystem.
// Returns true if at least one phase ran successfully.
func (p *Provisioner) Run(ctx context.Context) (bool, error) {
	deblobOK := false
	shrinkOK := false
	convertOK := false

	backend, err := p.createBackend(ctx)
	if err != nil {
		return false, err
	}

	// Phase 0: filesystem conversion (e.g. btrfs → ext4)
	if p.convertTo != "" {
		if err := backend.ConvertTo(ctx, p.convertTo); err != nil {
			return false, err
		}
		p.fsType = p.convertTo
		p.convertTo = ""
		convertOK = true
		slog.Info("Filesystem converted")
	}

	// Phase 1: deblob + fstab fix
	if p.deblob {
		if err := backend.Deblob(ctx, nil); err != nil {
			return false, err
		}
		if err := backend.Run(ctx); err != nil {
			return false, err
		}
		p.deblob = false
		deblobOK = true
	}

	// Phase 2: shrink
	if p.shrink {
		if err := backend.Shrink(ctx); err != nil {
			return false, err
		}
		if err := backend.Run(ctx); err != nil {
			return false, err
		}
		p.shrink = false
		shrinkOK = true
	}

	return convertOK || deblobOK || shrinkOK, nil
}

// ExtractViaBackend extracts a root partition from a raw disk image.
// Uses the selected backend's ExtractPartition method.
func ExtractViaBackend(
	ctx context.Context,
	rawPath, outputPath string,
	partition int,
	disabledDetectors []string,
	provisionerType provisioner.ProvisionerType,
) (result string, err error) {
	// Wrap non-DomainError as ImageError
	defer func() {
		if err != nil {
			var de *errs.DomainError
			if !errors.As(err, &de) {
				err = errs.New(errs.CodeImageError, err.Error())
			}
		}
	}()

	cacheDir, err := infra.GetCacheDir()
	if err != nil {
		return "", fmt.Errorf("resolve cache dir: %w", err)
	}

	backend, err := provisioner.NewBackend(ctx, provisioner.BackendOpts{
		RootfsPath:      rawPath,
		FsType:          "ext4", // placeholder — backends detect fs from the image
		CacheDir:        cacheDir,
		ProvisionerType: provisionerType,
		RootUID:         0,
		RootGID:         0,
		UserUID:         1000,
		UserGID:         1000,
	})
	if err != nil {
		return "", err
	}

	return backend.ExtractPartition(ctx, rawPath, outputPath, partition, disabledDetectors)
}
